alf/metrics/metrics.py

```diff
@@ -266,6 +279,68 @@ def _extract_metric_values(self, time_step):
             return [reward[..., i] for i in range(reward.shape[-1])]
 
 
+class AverageDiscountedEpisodicReturnMetric(AverageEpisodicSumMetric):
+    """Metric for computing the average discounted episodic return."""
+
+    def __init__(self,
+                 name='AverageDiscountedEpisodicReturnMetric',
+                 prefix='Metrics',
+                 reward_shape=(),
+                 dtype=torch.float32,
+                 discount=0.99,
+                 batch_size=1,
+                 buffer_size=10):
+        if reward_shape == ():
+            example_metric_value = torch.zeros((), device='cpu')
+        else:
+            example_metric_value = torch.zeros(
+                reward_shape, device='cpu').reshape(-1)
+            example_metric_value = list(example_metric_value)
+
+        self._discount = discount
+        self._accumulated_discount = torch.ones(batch_size, device='cpu')
+
+        super(AverageDiscountedEpisodicReturnMetric, self).__init__(
+            name=name,
+            dtype=dtype,
+            prefix=prefix,
+            batch_size=batch_size,
+            buffer_size=buffer_size,
+            example_metric_value=example_metric_value)
+
+    def _extract_metric_values(self, time_step):
+        """Accumulate discounted immediate rewards to get discounted episodic
+        return. It also updates the accumulated discount: multiply by
+        ``discount`` if not at episode beginning; and reset to one otherwise.
+        """
+        ndim = time_step.step_type.ndim
+        if time_step.reward.ndim == ndim:
+            return time_step.reward * self._accumulated_discount
+        else:
+            reward = time_step.reward.reshape(*time_step.step_type.shape, -1)
+            return [
+                reward[..., i] * self._accumulated_discount
+                for i in range(reward.shape[-1])
+            ]
+
+    def _after_update(self, time_step):
```

Comment:
I don't see the need to introduce another hook. Why don't you do this inside _extract_metric_values()?

Reply:
Originally to make different functionalities more isolated. I can merge remove and merge.

Reply:
Removed this function.